### ISSUE TITLE
[Feature Request]: Code completion using local LLM

### ISSUE THREAD

**Issue by maxguru** [eranif/codelite#3360]
### Describe your idea in details

It has now become possible to run a local LLMs with a reasonable speed. There is a large selection of models available that can be used. There are many apps out there already that make it easy to run models locally:
* [Ollama](https://github.com/ollama/ollama)
* [gpt4all](https://github.com/nomic-ai/gpt4all)
* [LM Studio](https://lmstudio.ai/)

It may not even be necessary to require local LLMs to be used. It would be possible to have support for external LLM APIs, should the user decide to go that route.

I suggest a plugin for CodeLite similar to GitHub Copilot's Visual Studio Code extension.

Possible Features:
* Model / API selection and configuration
* Code completion in code editor
* AI Chat in IDE widget
* Retrieval-augmented generation (RAG) using project files
* Autonomous task completion using function calling

This feature is critical for CodeLite. It would significantly enhance the coding experience, reducing development time and increasing productivity. It would allow CodeLite to compete with other IDEs that either have this feature already or will have it in the near future.

**Comment by genpfault** [comment]
Might be able to bodge something together by using/extending the existing [Language Server Protocol (LSP)](https://en.wikipedia.org/wiki/Language_Server_Protocol) support with something like [LSP-AI](https://github.com/SilasMarvin/lsp-ai/).

**Comment by eranif** [comment]
wow, thats really nice.
I already added support for local LLM chat integration in latest master build.
But this looks like something that we must have

**Comment by genpfault** [comment]
> I already added support for local LLM chat integration in latest master build.

Neat!  Looks like that lives in the [ChatAI plugin](https://github.com/eranif/codelite/tree/17.12.1/ChatAI), as of at least [`17.12.1`](https://github.com/eranif/codelite/releases/tag/17.12.1).

**Comment by maxguru** [comment]
This is great, however, the chat feature on its own isn't terribly useful. One can always open the same thing in a terminal window outside Codelite. What would make LLMs useful is the integration with the IDE, such as:
- RAG: provide context information (project files and code) to the model so it can provide relevant responses.
- Function calling:
  - Run commands on the project files (such as grep/etc) to gain information to answer questions.
  - Make edits to the project files (with user permission).
  - Run commands to install dependencies, etc.

Of course, one of the killer apps for LLMs is code completion. This feature isn't related to the chat. The editor window needs a feature that would show code suggestion (from the model) in the code being edited itself around where the cursor is and allow the user to accept or reject it by pressing some keyboard key combo.

Also, if you allow me to suggest more improvements, I would propose that the model selection and downloading be built into the settings interface. The features I describe above may require specific models and configurations to work well, so, it would be very helpful to have a preconfigured list for users to install by clicking a button. Additionally, recommendations can be made to the user regarding model choice, taking into account their machine's hardware and software limitations.

Many users aren't going to have access to hardware good enough to run a decent model. So, allowing outsourcing LLM calls to popular LLM APIs would be required. I guess my initial feature request was talking about a local LLM, however, it doesn't matter where the model runs in the grand scheme of things. The important thing is that Codelite attain the LLM integration features I described, otherwise it may become irrelevant in the future.

**Comment by maxguru** [comment]
By the way, wouldn't installing `/usr/bin/llama-cli` potentially create a conflict with the same binary installed by another package on the system?

**Comment by maxguru** [comment]
I like what devs behind [Continue](https://continue.dev) have done.

**Comment by eranif** [comment]
The way I see it, I think that eventually, LLM code completion will be integrated into an LSP server (similar to the work done here: https://github.com/huggingface/llm-ls)

**Comment by eranif** [comment]
FYI: I am re-writing the `ChatAI` to become more useful. I am building it around ollama where user can add MCP servers. In addition, CodeLite itself is an MCP server so it can register methods with the LLM for execution. For example:

- Write file to disk
- Open file for editing
- Read file content
- Read the compiler output

(many more will follow, but this will consume all our context window, so we need to be careful of how much we extend it).

With the above list, one can instruct the model with the following prompt:

```
Write me a small HTTP server in Rust. Write the server content in src/server.rs, in addition write the Cargo.toml file for building the program. When done, open both files for editing
```

The above will do exactly as the prompt describes. 

Another nice feature is:

```
explain the build error I got from the compiler
```

The LLM will read the build output from the build view and will analyze and explain how to fix this.

I will share a small video soon

**Comment by eranif** [comment]
![Image](https://github.com/user-attachments/assets/91e6cd46-4d6f-4d89-932a-19240115ea85)

**Comment by maxguru** [comment]
This is excellent! I'm looking forward to using it. As I recall, ollama has a setting for context window size and also this depends on the model. The context and tools can also be pruned by using various methods.

**Comment by eranif** [comment]
It can be configured via the options property in the config of the plugin. I added some more nice features (e.g. the plugin exposes API for other plugins to use, so the git plugin now has a "generate commit message" button, I will upload a small video for this as well.

**Comment by eranif** [comment]
![Image](https://github.com/user-attachments/assets/3900b51d-7e10-456c-a45c-809fa4469e27)

**Comment by maxguru** [comment]
It would be great for plugins to be able to add to Codelite MCP server tools, such as in the case of the git plugin:
- git status
- git log
- git show
- git commit

**Comment by eranif** [comment]
CodeLite itself is an MCP server, so it provides the model access to some of its functionality. For example, lets say you compile a code and got an error, you can open the chat box (`Ctrl-Shift-H`) and type:

```text
Explain the error I got from the compiler.
```

The model will active the `Read the compiler build output` (https://github.com/eranif/codelite/blob/master/ChatAI/Tools.cpp#L93) tool to fetch the compiler output and it will analyze and explain how to solve the problem. In addition, you can write you own tools (the easiest way is to use `python` and `FastMCP`) that provide extra functionality to the model and configure them in the configuration setup of the plugin.

The list of tools exposed by CodeLite are (notice that I asked the model to list them for me ;):

```markdown
**eran**:
List the tools available for you. List the required params for each tool. Use bullets.

**Assistant**:
The following tools are available for me:

- Get the text of the active tab inside the editor
  - No required parameters

- Open a file in an editor
  - filepath (string): The path of the file to open inside the editor.

- Read file from the file system
  - filepath (string): The path of the file to read.

- Read the compiler build output
  - No required parameters

- Write file content to disk at a given path
  - file_content (string): The content of the file to be written to the disk.
  - filepath (string): The path file path.
```

**Comment by maxguru** [comment]
Please excuse my ignorance of the Codelite code base if I get anything wrong.

What I was suggesting is something like `for(auto& plugin : plugins) plugin.PopulateBuildInFunctions(function_table);` in [`OllamaClient::WorkerThreadMain()`](https://github.com/eranif/codelite/blob/f32e41725860c60d3cb348280d1bce570a4d9727/ChatAI/OllamaClient.cpp#L98) in addition to `ollama::PopulateBuildInFunctions(function_table);`. This implies a change to `IPlugin` interface to add `PopulateBuildInFunctions` method. This would allow plugins to add more tools to what is already built in with [`ollama::PopulateBuildInFunctions(FunctionTable& table)`](https://github.com/eranif/codelite/blob/f32e41725860c60d3cb348280d1bce570a4d9727/ChatAI/Tools.cpp#L68C6-L68C52).

So, the git plugin might have it's own `PopulateBuildInFunctions` method which will add more git-related tools to `function_table`.

Another idea might be to add an `ExecuteCommand` tool, in which case these extra git tools that I am talking about might not even be necessary.

**Comment by eranif** [comment]
You are not mistaken. Since CodeLite uses "plugin" based architecture, it requires another level of abstraction to implement that.
A plugin (e.g. "git") is not aware of another plugin existence (e.g. the "git" plugin is not aware that the "ChatAI" plugin is loaded - and it shouldn't).  So any registered commands are going through an SDK class `LLMManager` (which all plugins are aware of its existence and it is always available) . 

This class will be enhanced with new methods:

```c++
bool LLMManager::RegisterFunction(LLMFunction func)
{
    std::string name = func.name;
    return m_functions.insert({name, std::move(func)}).second;
}

void LLMManager::UnregisterFunction(const std::string& funcname)
{
    if (m_functions.count(funcname) != 0) {
        // Erase it from the local cache.
        m_functions.erase(funcname);
    }

    // In case someone already consumed it from the LLMManager, notify that this function should be unregistered.
    clLLMEvent unregister_event{wxEVT_LLM_UNREGISTER_FUNC};
    unregister_event.SetStringRaw(funcname);
    unregister_event.SetString(funcname);
    EventNotifier::Get()->ProcessEvent(unregister_event);
}

void LLMManager::ConsumeFunctions(std::function<void(LLMFunction)> cb)
{
    for (auto& [_, f] : m_functions) {
        cb(std::move(f));
    }
    m_functions.clear();
}
```

So any plugin that wants to register a method, simply calls: `LLMManager::GetInstance().RegisterFunction(..)`
If the ChatAI plugin is loaded, it will be notified on any new function and it will register it into its `FunctionTable`.

**Comment by eranif** [comment]
This is now implemented. The git plugin now also registers a "git log" function (between 2 commits).

**Comment by eranif** [comment]
Going to close this issue. CodeLite now supports this and much more. I embedded a LUA scripting engine inside CodeLite where you can add custom context menus to CodeLite that will interact with the built-in LLM manager. This opens the door for a whole world of possibilities.

https://docs.codelite.org/LUA/lua/#llm-functions_1

The AI docs:
https://docs.codelite.org/AI/ai/
